martypy/ClientMV2.py

```diff
@@ -407,6 +408,95 @@ def is_conn_ready(self) -> bool:
             return False
         return self._initComplete and (self.lastSubscribedMsgTime is not None)
 
+    def _valid_addon(self, add_on: str) -> bool:
+        disco = {"00000087","00000088","00000089"}
+        for attached_add_on in self.get_add_ons_status().values():
+            if attached_add_on['name'] == add_on:
+                if attached_add_on['whoAmITypeCode'] in disco:
+                    return True
+                else:
+                    raise MartyCommandException("The add on name passed in is not a valid disco add on. Please check the add on name in the scratch app -> configure -> add ons")
+        raise MartyCommandException("The add on name passed in is not a valid add on. Please check the add on name in the scratch app -> configure -> add ons")
+
+    def _valid_hex(self, color_hex: str) -> bool:
+        hex_pattern = re.compile("^([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$")
+        return bool(hex_pattern.match(color_hex))
+
+    def disco_off(self, add_on: str) -> bool:
+        if self._valid_addon(add_on):
+            return self.add_on_query(add_on, bytes.fromhex('01'), 0)
+
+    def disco_pattern(self, pattern: str, add_on: str) -> bool:
+        if pattern == 1:
+                pattern = '10'
+        elif pattern == 2:
+            pattern = '11'
+        else:
+            raise Exception("Pattern must be 1 or 2")
+        if self._valid_addon(add_on):
+            return self.add_on_query(add_on, bytes.fromhex(pattern), 0)
+
+    def _region_to_bytes(self, region: Union[str, int]) -> bytes:
+        if region == 'all':
+            region = (2,)
+        else:
+            region = (4, region)
+        return bytes(region)
+
+    def _downscale_color(self, color: Union[tuple, bytes]):
+        return bytes(c//25 for c in color)
+
+    def _hex_to_bytes(self, color_hex: str, region: Union[str, int]) -> bytes:
+        color_hex = color_hex.lstrip('#')
+        if self._valid_hex(color_hex):
+            color_bytes = bytes.fromhex(color_hex)
+        else:
+            raise MartyCommandException("Color specified is not a valid hex code or default color")
```

Comment:
Can you include the "specified color" in the error message? Imagine someone uses this in a more complicated program. If their colour string gets messed up somewhere along the way, they would have to go back and add a print statement just to see what it is that MartyPy is complaining about exactly. We have the information readily available tho, so we can save them that step. For example:
```suggestion
            raise MartyCommandException(f'The string "{color_hex_original}" is not a valid hex color code or default color')
```
(You'd need to rename `color_hex` to retain access to the original argument passed in. (I'm not advocating for calling it `color_hex_original` though. 😄))

I think a similar thing applies to most other exceptions you raise. Basically, if you are complaining about what someone passed in, it is usually a good idea to show them what they passed in.